### PR TITLE
Retire email address field

### LINKED ISSUE #6400: User `Username or email` instead of requiring username and email in the same time
```
docker push foo/bar
The push refers to a repository [foo/bar] (len: 1)
Sending image list

Please login prior to push:
Username: foo
Password: 
Email: 
```

it's kinda annoying to enter email after entering username and password, I am not registering a new account here, why I have to enter my email? Although the .cfg file can save the logging information for me, I have multiple accounts for different usage, need to login different accounts sometimes. I think it's absolute not necessary to ask user both username and email with password.

I suggest what it should be is something like

```
Username or email: 
Password:
```

This is a more reasonable design.

### PR DESCRIPTION
**Pull request by tt**
I'm working on a third-party registry and it confused me that `docker login` will prompt for an email address which isn't even used for authentication.

This change will drop the unused email address while keeping support for reading configuration files that include the email address (plenty of the current tests assert that).

closes https://github.com/docker/docker/issues/6400

### CODE REVIEW

**Comment by tt** [comment]
The experimental build failure seems to be unrelated although I'm not sure if this is expected.

**Comment by aaronlehmann** [comment]
Distribution is +1 on this.

**Comment by icecrime** [comment]
Fine by me if distribution is ok for it.

- @tt do you please mind squashing your commits into 1?
- @jlhawn ping for `AuthConfig` impact!

**Comment by jlhawn** [comment]
Does this mean we're also retiring the Registration feature of `docker login`? I'm okay with not storing it in the config file (it's never used again anyway). I'm even more okay with **not** having a user signup feature in the CLI but there are higher powers involved that have kept that feature alive when there have been several attempts in the past to remove it.

**Comment by thaJeztah** [comment]
If we decide to retire registration from the CLI, a note needs to be added to https://github.com/docker/docker/blob/master/docs/misc/deprecated.md

**Comment by thaJeztah** [comment]
fwiw, I'm +1 on retiring it, it has come up several times in the past :-)

**Comment by HuKeping** [comment]
+1 for retiring the `email`

File: api/client/login.go [hunk 1]
```diff
@@ -70,7 +69,7 @@ func (cli *DockerCli) CmdLogin(args ...string) error {
 		}
 	}
 	// Assume that a different username means they may not want to use
-	// the password or email from the config file, so prompt them
+	// the password from the config file, so prompt them
```
Comment (HuKeping) [inline api/client/login.go]: `them` -> `it`?

**Comment by SvenDowideit** [comment]
There are/where also instructions on how to create a new account on hub using `docker login` that would need to be removed - see `/docs/userguide/dockerrepos.md`

**Comment by aaronlehmann** [comment]
See also https://github.com/docker/distribution/issues/308

**Comment by stevvooe** [comment]
:+1: for the idea. Automated registration is always confusing for users. I said most of my opinion on this in https://github.com/docker/distribution/issues/308#issuecomment-88244735.

**Comment by thaJeztah** [comment]
Oh, think we can add: closes https://github.com/docker/docker/issues/12991

**Comment by thaJeztah** [comment]
@aaronlehmann @icecrime is this ready for docs review? I see approvals, but no LGTM's yet

File: api/client/login.go (hunk 1 above)
Comment (endophage) [inline api/client/login.go]: "them" is correct, it's referencing the user.

**Comment by endophage** [comment]
+1 for change.

Most of the test failures look unrelated but this one (the last failure jenkins lists) looks like it may be related, just because it fails on a login attempt and I don't see any updates to the integration tests.

```FAIL: github.com/docker/docker/integration-cli/_test/_obj_test/docker_cli_v2_only.go:97: DockerRegistrySuite.TestV1

[d64837959] waiting for daemon to start
[d64837959] waiting for daemon to start
[d64837959] waiting for daemon to start
[d64837959] waiting for daemon to start
[d64837959] waiting for daemon to start
[d64837959] waiting for daemon to start
[d64837959] waiting for daemon to start
[d64837959] daemon started
github.com/docker/docker/integration-cli/_test/_obj_test/docker_cli_v2_only.go:175:
    ...open github.com/docker/docker/integration-cli/_test/_obj_test/docker_cli_v2_only.go: no such file or directory
... Error: Expected v1 login attempt

github.com/docker/docker/integration-cli/_test/_obj_test/docker_cli_v2_only.go:184:
    ...open github.com/docker/docker/integration-cli/_test/_obj_test/docker_cli_v2_only.go: no such file or directory
... Error: Not all endpoints contacted after push2 0

[d64837959] exiting daemon```

**Comment by jfrazelle** [comment]
moved to docs review, can you please squash the commits

**Comment by thaJeztah** [comment]
@jfrazelle given that this is an API change, should a "versioned" struct be added for older API versions?
See https://github.com/docker/docker/pull/16564.

Also when searching for docs that need an update, I noticed this [registry/auth.go#L72](https://github.com/docker/docker/blob/634a848b8e3bdd8aed834559f3b2e0dfc7f5ae3a/registry/auth.go#L72), and wasn't sure if that's still relevant?


Thanks @tt! Some pointers for docs and other changes that I think are needed:

I noticed this also needs updates in the completion scripts;

- Bash: https://github.com/docker/docker/blob/master/contrib/completion/bash/docker#L915-L927
- zsh: https://github.com/docker/docker/blob/master/contrib/completion/zsh/_docker#L578-L584
- fish: https://github.com/docker/docker/blob/master/contrib/completion/fish/docker.fish#L226

(we can do that as a follow-up PR if you want, then I'll ping the maintainers for those scripts)



Looked up docs that mention this, and (I hope I have them all) came up with:

- We need a mention in [Deprecated Features](https://github.com/docker/docker/blob/master/docs/misc/deprecated.md)
- This contains a change in the API, so needs a mention in the [v1.21 API changes](https://github.com/docker/docker/blob/master/docs/reference/api/docker_remote_api.md#v121-api-changes)
- [/man/docker-login.1.md](https://github.com/docker/docker/blob/634a848b8e3bdd8aed834559f3b2e0dfc7f5ae3a/man/docker-login.1.md) (mentions the -e/--email option and the description contains "Register or login")
- same for [/docs/reference/commandline/login.md](https://github.com/docker/docker/blob/80469f8180506b8e3a04e63766b15caddffa9726/docs/reference/commandline/login.md)
- [api/docker_remote_api_v1.21.md](https://github.com/docker/docker/blob/246072e8c5184c78edc9f40435987a7ba26baaa8/docs/reference/api/docker_remote_api_v1.21.md#check-auth-configuration) mentions "Get the default username and email", and shows the e-mail in the output

**Comment by thaJeztah** [comment]
Oh, almost forgot the document @SvenDowideit mentioned; https://github.com/docker/docker/blob/master/docs/userguide/dockerrepos.md

**Comment by thaJeztah** [comment]
ping @tt could you address the documentation changes?

**Comment by vdemeester** [comment]
@tt Don't forget to squash your commits too :wink:

**Comment by thaJeztah** [comment]
ping @kencochrane, I just discussed this with Tibor and Michael, and we may need the input of the Hub team on this change, to check if you agree with making this change

**Comment by icecrime** [comment]
I don't understand why this is in docs review, there are no code LGTM.

**Comment by thaJeztah** [comment]
Had a quick chat with @kencochran. There's still a discussion going on internally for Docker Hub; overall they're not against this, but there are changes needed on their side to facilitate this, also for backward compatibility with older clients. Since 1.10 is still roughly two months away, that may be time enough for them, but I need to check further with some other people and will leave a comment if I know more

**Comment by friism** [comment]
This seems to also be a problem for Google: https://cloud.google.com/container-registry/docs/auth#using_an_access_token

```
Email	Unused, must be a well-formed email address (e.g 1234@5678.com).
```

**Comment by jfrazelle** [comment]
can you squash your commits

**Comment by thaJeztah** [comment]
ping @bfirsh I've assigned you to this issue, following the e-mail discussion that you'll be coordinating this with Docker Hub team :smile:

**Comment by thaJeztah** [comment]
And apologies to @tt that this is taking so long, thanks for being patient with us! :heart:

**Comment by stevvooe** [comment]
> This seems to also be a problem for Google: https://cloud.google.com/container-registry/docs/auth#using_an_access_token
> Email   Unused, must be a well-formed email address (e.g 1234@5678.com).

@mattmorr Any way to get this relaxed?

**Comment by mattmoor** [comment]
@stevvooe We only have that because it's required by `docker login`.  Not sure what you think needs to be relaxed?

**Comment by stevvooe** [comment]
@mattmoor The wording there makes it sound like if the Email field is missing, the login will be refused. If we remove the field from the docker client, will one still be able to login to GCS?

**Comment by mattmoor** [comment]
@stevvooe Absolutely, we only put garbage strings in here today.  We can qualify this as important for certain versions of the client if you plan to remove it, thanks for the heads up.

**Comment by tiborvass** [comment]
Ping @bfirsh

**Comment by bfirsh** [comment]
Currently putting together a proposal for what we can do here – will give an update in a day or two.

**Comment by dmp42** [comment]
@bfirsh bump

**Comment by bfirsh** [comment]
Still working with Hub team to figure out something workable – will have something soon!

**Comment by bfirsh** [comment]
I've written up an improved design here with impact to both Engine and Hub: https://github.com/docker/docker/issues/18668

Comments welcome!

**Comment by tiborvass** [comment]
@bfirsh when can we resume work on this PR?

**Comment by bfirsh** [comment]
@tiborvass This is blocked on the Hub not requiring email addresses, which is being worked on at the moment.

**Comment by vdemeester** [comment]
Should we close it and re-open/carry once it's done on the Hub side ? :wink:

**Comment by jfrazelle** [comment]
that sounds like a good idea.

**Comment by tiborvass** [comment]
Ping @bfirsh any updates?

**Comment by bfirsh** [comment]
@tiborvass Afraid not – Hub have a lot on their plate. :(

Will let you know when I have an update from them.

**Comment by stevvooe** [comment]
@bfirsh Let's get an estimate together. No reason to keep pushing this back.

**Comment by JonathonReinhart** [comment]
@jfrazelle Is it really a good practice to close issues just because they are blocked? Most everywhere else, closing an issue means it's resolved, usually "fixed" or "won't fix".

**Comment by vdemeester** [comment]
@JonathonReinhart first it's a pull request, not an issue ; the related issue is [there](https://github.com/docker/docker/issues/18668) and is still open :wink:. This pull-request has been closed because it stalled and we are waiting for changes from upstream :stuck_out_tongue_closed_eyes:.

**Comment by thaJeztah** [comment]
@JonathonReinhart this is a -pull request-, not an issue. The pull request cannot be worked on pending the dependency on hub changes, so there's no need to keep it open for now (related issue is https://github.com/docker/docker/issues/6400)

**Comment by JonathonReinhart** [comment]
@jfrazelle @vdemeester @thaJeztah Sorry guys, too early in the morning :-)  I'll remove my comments  in a few minutes as to not clutter up the conversation here.

**Pull request closed without merging**